packages/block-editor/src/store/reducer.js

```diff
@@ -830,84 +812,98 @@ export const blocks = pipe(
 		return state;
 	},
 
-	attributes( state = {}, action ) {
+	attributes( state = new Map(), action ) {
 		switch ( action.type ) {
 			case 'RECEIVE_BLOCKS':
-			case 'INSERT_BLOCKS':
-				return {
-					...state,
-					...getFlattenedBlockAttributes( action.blocks ),
-				};
+			case 'INSERT_BLOCKS': {
+				const newState = new Map( state );
```

Comment:
@youknowriad I'm inclined to add a comment documenting the choice of operations for a tangible performance hot-path optimization here. It'd be easy to lose track of it in a future refactor

Reply:
There are [reasons](https://redux.js.org/style-guide/#do-not-put-non-serializable-values-in-state-or-actions) why we normally don't put `Map`s or other non-serializable values in the Redux store. It's no longer showing the state in the Redux devtools for me after this PR.

> I tried "Immer" as it's also a popular library in Redux land, but it turned out be way less performant.

I wonder if there's any benchmark on this? Seems like Immer should also be [performant](https://immerjs.github.io/immer/performance) in most cases. Were we testing it in the production build?

(Side note: adopting Immer should also help us when we want to integrate [`yjs`](https://github.com/yjs/yjs) more closely into the system for collaborative editing.)

I think we should at least bring back the ability to debug values in the Redux devtools. Maybe the [`serialize`](https://github.com/zalmoxisus/redux-devtools-extension/blob/master/docs/API/Arguments.md#serialize) option would help?

Reply:
Is there any specific reason why Maps should be avoided, other than Redux DevTools not working with them (which to me sounds like a Redux DevTools problem)?

The link only goes on to mention:
> "It also ensures that the UI will update as expected."

I'm not sure which UI this refers to, nor why it would stop updating as expected.

Reply:
> I wonder if there's any benchmark on this? 

It's a benchmark I did on my own using the same test included in this PR. A refactor using immer is actually very simple but it was three times worse in terms of performance.

Reply:
Good call on the "serialize" thing, I'll be fixing that in a follow-up PR.

Reply:
Follow-up here https://github.com/WordPress/gutenberg/pull/46282